tensornetwork/quantum/quantum.py

```diff
@@ -0,0 +1,374 @@
+
+  def norm(self):
+    """The norm of the operator.
+    This is the 2-norm (also known as the Frobenius or Hilbert-Schmidt norm).
+    """
+    return (self.adjoint() @ self).trace()
+
+  def partial_trace(self, subsystems_to_trace_out: Collection[int]):
+    """The partial trace of the operator.
+
+    Subsystems to trace out are supplied as indices, so that dangling edges
+    are connected to eachother as:
+      `out_edges[i] ^ in_edges[i] for i in subsystems_to_trace_out`
+
+    This does not modify the original network. The original ordering of the
+    remaining subsystems is maintained.
+
+    Args:
+      subsystems_to_trace_out: Indices of subsystems to trace out.
+    Returns:
+      A new QuOperator or QuScalar representing the result.
+    """
+    check_spaces(self.in_edges, self.out_edges)
+
+    out_edges_trace = [self.out_edges[i] for i in subsystems_to_trace_out]
+    in_edges_trace = [self.in_edges[i] for i in subsystems_to_trace_out]
+
+    nodes_dict, edge_dict = copy(self.nodes, False)
+    for (e1, e2) in zip(out_edges_trace, in_edges_trace):
+      edge_dict[e1] = edge_dict[e1] ^ edge_dict[e2]
+
+    # get leftover edges in the original order
+    out_edges_trace = set(out_edges_trace)
+    in_edges_trace = set(in_edges_trace)
+    out_edges = [edge_dict[e] for e in self.out_edges
+                 if e not in out_edges_trace]
+    in_edges = [edge_dict[e] for e in self.in_edges
+                if e not in in_edges_trace]
+    ref_nodes = [n for _, n in nodes_dict.items()]
+    ignore_edges = [edge_dict[e] for e in self.ignore_edges]
+
+    return quantum_constructor(out_edges, in_edges, ref_nodes, ignore_edges)
+
+  def __matmul__(self, other: "QuOperator"):
+    check_spaces(self.in_edges, other.out_edges)
+
+    # Copy all nodes involved in the two operators.
+    # We must do this separately for self and other, in case self and other
+    # are defined via the same network components (e.g. if self === other).
+    nodes_dict1, edges_dict1 = copy(self.nodes, False)
+    nodes_dict2, edges_dict2 = copy(other.nodes, False)
+
+    # connect edges to create network for the result
+    for (e1, e2) in zip(self.in_edges, other.out_edges):
+      _ = edges_dict1[e1] ^ edges_dict2[e2]
+
+    in_edges = [edges_dict2[e] for e in other.in_edges]
+    out_edges = [edges_dict1[e] for e in self.out_edges]
+    ref_nodes = ([n for _, n in nodes_dict1.items()] +
+                 [n for _, n in nodes_dict2.items()])
+    ignore_edges = ([edges_dict1[e] for e in self.ignore_edges] +
+                    [edges_dict2[e] for e in other.ignore_edges])
+
+    return quantum_constructor(out_edges, in_edges, ref_nodes, ignore_edges)
+
+  def __mul__(self, other: "QuOperator"):
+    if self.is_scalar or other.is_scalar:
+      return self @ other
+    raise ValueError("Elementwise multiplication is only supported if at "
+                     "least one of the arguments is a scalar.")
+
+  def tensor_prod(self, other: "QuOperator"):
```

Comment:
in the rest of the code this is called the `outer_product`. should we do that here too?

Reply:
Yeah, I wondered about this too...
I chose `tensor_prod` because this is the common parlance in QM. We could have both `tensor_product` and `outer_product` mean the same thing?